Added parse-tree-whitequark CLI option.

iliabylich opened the pull request:
<!-- (optional) Explain your change, focusing on the details of the solution. This is a great place to call out user-visible changes. -->

+ Added `parse-tree-ruby` formatting option. Prints what `whitequark/parser` prints (i.e. the return value of the `AST::Node#inspect` method)
+ Added test for the CLI
+ I didn't write any C++ for many years, and I don't know anything about sorbet internals, so feel free to point to / suggest any changes.

### Motivation
<!-- Why make this change? Describe the problem, not the solution. This can also be a link to an issue. -->

1. This format is more readable (that's personal though)
2. This format simplifies backporting tests from `whitequark/parser` (all we need is to run tets from parser with custom assertions that record inputs/outputs)
3. It simplifies debugging the difference between currently used parser and `whitequark/parser` by comparing `sorbet -p parse-tree-ruby -e <code>` and `ruby-parse -e 'code'`.

### Test plan
<!-- If you did not write tests for this change, replace the message below explaining why not. Why we should be confident this change is correct? If you changed the website, please include a screenshot of the proposed changes. -->

See included automated tests.

DarkDimius commented on parser/tools/generate_ast.cc:
```diff
@@ -375,6 +483,60 @@ void emitNodeClassfile(ostream &out, NodeDef &node) {
+    out << "    fmt::format_to(buf, \"s(:" << rubyNodeNameForInternalNodeName(node.name) << "\");" << '\n';
+    // Generate fields
+    for (auto &arg : node.fields) {
+        if (arg.type == Loc) {
```
This implicitly requires us to order fields in the same way how whitequark does.
Is it already true today?
I have minor fud that it might be hard to keep them in sync, but I'm happy to wait till this FUD becomes reality.
In the mean time, I guess this is a debugging tool that will make your life better.

iliabylich replied:
> This implicitly requires us to order fields in the same way how whitequark does.

Yes, and this order feels right to me (and so I think it makes sense to "backport" it as well). For example, an `If` node has two children: truthy and falsey branches. It sounds correct to declare them as
```
{"then_", Node}, {"else_", Node}
```

> Is it already true today?

Yes, mostly. However there's some difference in the AST format: sorbet has an `assign` node, `int` node has (obviously) a string as the only child, plain string literals without interpolation are always wrapped into `dstr` and maybe more, I didn't have a chance to look into all of them.

DarkDimius commented on main/options/options.cc:
```diff
@@ -27,6 +27,7 @@ struct PrintOptions {
 const vector<PrintOptions> print_options({
     {"parse-tree", &Printers::ParseTree},
     {"parse-tree-json", &Printers::ParseTreeJson},
+    {"parse-tree-ruby", &Printers::ParseTreeRuby},
```
do other tools other than whitequark use this format?
If not, WDYT about `parse-tree-whitequark-like` The `like` there is for "best effort"

DarkDimius commented:
Kicking off ci

jez commented on parser/tools/generate_ast.cc:
```diff
@@ -375,6 +483,60 @@ void emitNodeClassfile(ostream &out, NodeDef &node) {
+    // toRuby
+    out << "  std::string " << node.name << "::toRuby(const core::GlobalState &gs, int tabs) {" << '\n'
+        << "    fmt::memory_buffer buf;" << '\n';
+    out << "    fmt::format_to(buf, \"s(:" << rubyNodeNameForInternalNodeName(node.name) << "\");" << '\n';
```
If we want it to be the case that this lookup always succeeds, why not just update `NodeDef` to have a third field which is the whitequark name? i.e., we'd have something like this:

```cpp
struct NodeDef {
    string name;
    string new_thing;
    vector<FieldDef> fields;
};

NodeDef nodes[] = {
    // alias bar foo
    {"Alias", "alias", vector<FieldDef>({{"from", Node}, {"to", Node}})},
    // ...
};

... << node.new_thing << ...
```

iliabylich replied:
I wanted to do it initially, but it feels dirty to me, and so I decided to extract it into external mapping. Different libraries/people may want to call nodes as they want, but that doesn't mean that all of this metadata is a part of the core logic.

jez commented on parser/tools/generate_ast.cc:
```diff
@@ -375,6 +483,60 @@ void emitNodeClassfile(ostream &out, NodeDef &node) {
+                break;
+            case Loc:
+                // quiet the compiler; we skip Loc fields above
+                abort();
```
Why have both a `switch` and an `if`? Could the first part of the switch just be

```cpp
switch (arg.type) {
    case Loc:
        continue;
    case Name:
    // ...
}
```

jez commented on main/options/options.cc:
```diff
@@ -27,6 +27,7 @@ struct PrintOptions {
 const vector<PrintOptions> print_options({
     {"parse-tree", &Printers::ParseTree},
     {"parse-tree-json", &Printers::ParseTreeJson},
+    {"parse-tree-ruby", &Printers::ParseTreeRuby},
```
(And then ideally update all the internal variable names to match)

iliabylich commented:
oh, I need to update a test for `--help`. Am I reading it properly? Is it the only failure?

DarkDimius commented:
Yes, `//test/cli:test_help` is the only test that failed

DarkDimius commented:
you can get it updated by running `./tools/scripts/update_exp_files.sh`

iliabylich commented:
@DarkDimius  @jez  I've addressed all comments, fixed tests, and added/improved a few comments in the nodes list. Also a few lines were too long, so I had to slightly reformat the nodes list. Could you take a look please?

jez commented:
Thanks!
